kore/src/Kore/Step/BaseStep.hs

```diff
@@ -887,3 +860,528 @@ unwrapPredicateVariables
+instantiated to the solution (see 'instantiateRule'). The initial conditions are
+not applied to the rule (see 'applyRule').
+
+ -}
+unifyRule
+    ::  ( Ord     (variable Object)
+        , Show    (variable Object)
+        , Unparse (variable Object)
+        , FreshVariable  variable
+        , SortedVariable variable
+        )
+    => MetadataTools Object StepperAttributes
+    -> UnificationProcedure Object
+    -> PredicateSubstitutionSimplifier Object
+    -> StepPatternSimplifier Object
+    -> BuiltinAndAxiomSimplifierMap Object
+
+    -> ExpandedPattern Object variable
+    -- ^ Initial configuration
+    -> RulePattern Object variable
+    -- ^ Rule
+    -> BranchT
+        (ExceptT (StepError Object variable) Simplifier)
+        (UnifiedRule variable)
+unifyRule
+    metadataTools
+    (UnificationProcedure unificationProcedure)
+    predicateSimplifier
+    patternSimplifier
+    axiomSimplifiers
+
+    initial@Predicated { term = initialTerm }
+    rule
+  = do
+    let
+        -- Rename free axiom variables to avoid free variables from the initial
+        -- configuration.
+        (_, rule') = RulePattern.refreshRulePattern configVariables rule
+          where
+            configVariables = ExpandedPattern.freeVariables initial
+        RulePattern { left = ruleLeft } = rule'
+    unifier <- unifyPatterns ruleLeft initialTerm
+    return (unifier $> rule')
+  where
+    unifyPatterns pat1 pat2 = do
+        (unifiers, _) <-
+            fromUnification
+            $ Monad.Trans.lift
+            $ unificationProcedure
+                metadataTools
+                predicateSimplifier
+                patternSimplifier
+                axiomSimplifiers
+                pat1
+                pat2
+        scatter unifiers
+
+{- | Instantiate the rule by applying the unification solution.
+
+The unification solution is normalized with the 'requires' clause from the
+unified rule. @instantiateRule@ fails if normalization fails. @instantiateRule@
+branches when the 'PredicateSubstitutionSimplifier' causes normalization to
+branch.
+
+ -}
+instantiateRule
+    ::  ( Ord     (variable Object)
+        , Show    (variable Object)
+        , Unparse (variable Object)
+        , FreshVariable  variable
+        , SortedVariable variable
+        )
+    => MetadataTools Object StepperAttributes
+    -> PredicateSubstitutionSimplifier Object
+    -> StepPatternSimplifier Object
+    -> BuiltinAndAxiomSimplifierMap Object
+
+    -> UnifiedRule variable
+    -- ^ Rule and unification solution
+    -> BranchT
+        (ExceptT (StepError Object variable) Simplifier)
+        (UnifiedRule variable)
+instantiateRule
+    metadataTools
+    predicateSimplifier
+    patternSimplifier
+    axiomSimplifiers
+
+    unified@Predicated { term = axiom@RulePattern { left, right, requires } }
+  = do
+    let unifier = unified { term = () }
+        merged = PredicateSubstitution.fromPredicate requires *> unifier
```

Comment:
It's late so I'm stopping the review here (sorry, I wanted to finish it today, but I can't).

I don't have time to look it up, so how does this thing work? *> is supposed to discard the first action, so I guess it must be something about some side effects? Anyway, is there a more obvious way of doing this?

Reply:
Had to look it up myself when reviewing. I *think* it uses the `Predicated/Monad` semantics, which is:

given a `p :: Predicated term` (with `term1 /\ predicate1 /\ substitution1` and a kleisli arrow `f :: term -> Predicated`
apply `f p` and obtain `term2 /\ predicate2 /\ substitution2`
return a Predicated like `term2 /\ (predicate1 /\ predicate2) /\ (substitituion1 <> substitution2)`

So basically, it is assumed that the function `f` only takes into account the `term` part of `p`.

Which in turn means that `p *> q` would mean:
`qterm /\ (ppredicate /\ qpredicate) /\ (psubstitution <> qsubstitution)`

Considering a `Predicate term` is `(term, predicate, substitution)`, this lines up with the tuple/pair Monad instance, which requires a `Monoid` instance on `fst`.

Reply:
@vladciobanu is right; the `Applicative` and `Monad` instances for `Predicated` combine the side conditions (predicates and substitutions) using `\and`. `p *> q` means "combine the predicates and substitutions of `p` and `q`, but take only the `term` from `q`". Actually, in this case the term of both `p` and `q` is `()`, so it wouldn't matter which one is taken.